### ISSUE TITLE
Incorrect overload parameter mismatch when typing decorator

### ISSUE THREAD

**Issue by max-kamps** [microsoft/pyright#11333]
**Describe the bug**
Trying to type a decorator, and I'm running into some seemingly spurious errors. I've tried to figure out what exactly causes this, to no avail.

The goal is to have the an optional `cls` parameter, and to have the return type vary based on whether it is present or absent. Mypy accepts the following code, but Pyright does not.

Apologies if this has already been reported, but I couldn't find any existing issue.

**Code **

[Pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQDClZZlARmcQFBcACRpCjS7ViwKKIDGuSjFwBtACoBdABSSyAZwBcsRMSXL6AenoAqesQAeMEJV2p8AXigAGAJRQAtAD49CAxVtLihQqAA6SJ5%2BEnIqahExCWJpOzkQQ3UtXQA5MBRiKBc8gpNzSxs7BxRnN09fRmZWDgN5eADDI39A5WVgsIioxPEpGXTMjR1uwygAHygSwuL84jKoCyhrW3tMGqK6-rCkcUnMTQWVw4HQ0XEAdzsEAJAsqfae%2Br93wyvrgZBiDAAK4gFDJVKyXCvCrbJxbOzuEJ-KAA4GgqAPShPUhI0K4qCaQGyWzQqAAIgA%2BhT4ZQqWSYQj8aiQWDJlwgA)

[Mypy playground](https://mypy-play.net/?mypy=latest&python=3.14&gist=41df62496abcd51282d3f8d5097ba4b1)

```py
from typing import overload, Callable

@overload
def decorator[T](cls: type[T], /, *, extra: int = 0) -> type[T]:
    ...

@overload
def decorator[T](cls: None = None, /, *, extra: int = 0) -> Callable[[type[T]], type[T]]:
    ...

def decorator[T](cls: type[T] | None = None, /, *, extra: int = 0):
    if cls is None:
        def wrapper(cls: type[T]) -> type[T]:
            return decorator(cls, extra=extra)
        return wrapper
    
    setattr(cls, "__extra__", extra)
    return cls
```

Pyright error message:
```py
Overloaded implementation is not consistent with signature of overload 2
  Function return type "(type[T@decorator]) -> type[T@decorator]" is incompatible with type "((cls: type[T@decorator]) -> type[T@decorator]) | type[T@decorator]"
    Type "(type[T@decorator]) -> type[T@decorator]" is not assignable to type "((cls: type[T@decorator]) -> type[T@decorator]) | type[T@decorator]"
      Type "(type[T@decorator]) -> type[T@decorator]" is not assignable to type "type[T@decorator]"
      Type "(type[T@decorator]) -> type[T@decorator]" is not assignable to type "(cls: type[T@decorator]) -> type[T@decorator]"
        Missing keyword parameter "cls"
          Position-only parameter mismatch; parameter "cls" is not position-only
          Position-only parameter mismatch; expected 1 but received 0  (reportInconsistentOverload)
```

**Comment by bschnurr** [comment]
adding a return type annotation fixes it . but i'm not sure why
```python
from typing import overload, Callable

@overload
def decorator[T](cls: type[T], /, *, extra: int = 0) -> type[T]:
    ...

@overload
def decorator[T](cls: None = None, /, *, extra: int = 0) -> Callable[[type[T]], type[T]]:
    ...

def decorator[T](cls: type[T] | None = None, /, *, extra: int = 0) -> type[T] | Callable[[type[T]], type[T]]:
    if cls is None:
        def wrapper(cls: type[T]) -> type[T]:
            return decorator(cls, extra=extra)
        return wrapper
    
    setattr(cls, "__extra__", extra)
    return cls
```

and maybe the Pylance function hover has some insight. 

<img width="967" height="633" alt="Image" src="https://github.com/user-attachments/assets/74076a39-3e03-4c0c-b394-c43714219fcc" />
